### PR TITLE
misc: use ibuf API to discard/allocate/consume.

### PR DESCRIPTION
**Pull request by nshy**
We have several places where we move ibuf.wpos pointer directly without using ibuf API. As a result we don't poison the area that is known to be unallocated.

Let's introduce `ibuf_discard` to discard given number of bytes from write end of the buffer.

Follow-up #7327

### CODE REVIEW

File: src/box/lua/net_box.c [hunk 1]
```diff
@@ -1142,7 +1142,7 @@ netbox_transport_communicate(struct netbox_transport *transport, size_t limit)
 						"Peer closed");
 				return -1;
 			} if (rc > 0) {
-				recv_buf->wpos += rc;
+				ibuf_alloc(recv_buf, rc);
```
Comment (locker) [inline src/box/lua/net_box.c:1145]: We advance `wpos` with `ibuf_alloc`, but `rpos` directly. Maybe, it's worth adding `ibuf_advance` helper that would poison memory before rpos?

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/lua/net_box.c (hunk 1 above)
Comment (nshy) [reply src/box/lua/net_box.c:1145]: Fixed.

```diff
diff --git a/src/box/lua/merger.c b/src/box/lua/merger.c
index adba94a58..b5add37ce 100644
--- a/src/box/lua/merger.c
+++ b/src/box/lua/merger.c
@@ -115,8 +115,8 @@ decode_header(struct ibuf *buf, size_t *len_p)
 static void
 encode_header(struct ibuf *output_buffer, uint32_t result_len)
 {
-       ibuf_reserve(output_buffer, mp_sizeof_array(result_len));
-       output_buffer->wpos = mp_encode_array(output_buffer->wpos, result_len);
+       mp_encode_array(xibuf_alloc(output_buffer, mp_sizeof_array(result_len)),
+                       result_len);
 }

 /**
@@ -1140,7 +1140,7 @@ encode_result_buffer(struct lua_State *L, struct merge_source *>
               merge_source_next(source, NULL, &tuple)) == 0 &&
               tuple != NULL) {
                uint32_t bsize = tuple_bsize(tuple);
-               void *wpos = ibuf_alloc(output_buffer, bsize);
+               void *wpos = xibuf_alloc(output_buffer, bsize);
                memcpy(wpos, tuple_data(tuple), bsize);
                result_len_offset += bsize;
                ++result_len;
```

File: src/box/lua/merger.c [hunk 2]
```diff
@@ -1140,9 +1140,8 @@ encode_result_buffer(struct lua_State *L, struct merge_source *source,
 	       merge_source_next(source, NULL, &tuple)) == 0 &&
 	       tuple != NULL) {
 		uint32_t bsize = tuple_bsize(tuple);
-		ibuf_reserve(output_buffer, bsize);
-		memcpy(output_buffer->wpos, tuple_data(tuple), bsize);
-		output_buffer->wpos += bsize;
+		void *wpos = ibuf_alloc(output_buffer, bsize);
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/merger.c:1143]: Fixed.

```diff
diff --git a/src/box/lua/merger.c b/src/box/lua/merger.c
index adba94a58..b5add37ce 100644
--- a/src/box/lua/merger.c
+++ b/src/box/lua/merger.c
@@ -115,8 +115,8 @@ decode_header(struct ibuf *buf, size_t *len_p)
 static void
 encode_header(struct ibuf *output_buffer, uint32_t result_len)
 {
-       ibuf_reserve(output_buffer, mp_sizeof_array(result_len));
-       output_buffer->wpos = mp_encode_array(output_buffer->wpos, result_len);
+       mp_encode_array(xibuf_alloc(output_buffer, mp_sizeof_array(result_len)),
+                       result_len);
 }

 /**
@@ -1140,7 +1140,7 @@ encode_result_buffer(struct lua_State *L, struct merge_source *>
               merge_source_next(source, NULL, &tuple)) == 0 &&
               tuple != NULL) {
                uint32_t bsize = tuple_bsize(tuple);
-               void *wpos = ibuf_alloc(output_buffer, bsize);
+               void *wpos = xibuf_alloc(output_buffer, bsize);
                memcpy(wpos, tuple_data(tuple), bsize);
                result_len_offset += bsize;
                ++result_len;
```

[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/merger.c:1143]: I also found more places where `wpos` is advanced. I forget to check for `wpos\s*=\s*' previously.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/lua/tuple.lua b/src/box/lua/tuple.lua
index a38a4f84e..584fe32ab 100644
--- a/src/box/lua/tuple.lua
+++ b/src/box/lua/tuple.lua
@@ -304,9 +304,7 @@ end
 local function tuple_to_msgpack(buf, tuple)
     assert(ffi.istype(tuple_t, tuple))
     local bsize = builtin.box_tuple_bsize(tuple)
-    buf:reserve(bsize)
-    builtin.box_tuple_to_buf(tuple, buf.wpos, bsize)
-    buf.wpos = buf.wpos + bsize
+    builtin.box_tuple_to_buf(tuple, buf:alloc(bsize), bsize)
 end
 
 local function tuple_bsize(tuple)
diff --git a/src/lua/httpc.lua b/src/lua/httpc.lua
index e694f1fc1..c24852fcf 100644
--- a/src/lua/httpc.lua
+++ b/src/lua/httpc.lua
@@ -441,7 +441,7 @@ local function io_read(self, opts, timeout)
             rbuf.rpos = rbuf.rpos + len
             return data
         else
-            rbuf.wpos = rbuf.wpos + res
+            rbuf:alloc(res)
             local len = check(self, chunk, delimiter)
             if len ~= nil then
                 self._errno = nil
diff --git a/src/lua/socket.lua b/src/lua/socket.lua
index 285814804..4f43e592a 100644
--- a/src/lua/socket.lua
+++ b/src/lua/socket.lua
@@ -723,7 +723,7 @@ local function read(self, limit, timeout, check, ...)
             rbuf.rpos = rbuf.rpos + len
             return data
         elseif res ~= nil then
-            rbuf.wpos = rbuf.wpos + res
+            rbuf:alloc(res)
             local len = check(self, limit, ...)
             if len ~= nil then
                 self._errno = nil
```

</details>

File: src/box/lua/net_box.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/net_box.c:1145]: Done.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index ee440832b..1da90c3da 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -1142,7 +1142,7 @@ netbox_transport_communicate(struct netbox_transport *transport, size_t limit)
 						"Peer closed");
 				return -1;
 			} if (rc > 0) {
-				ibuf_alloc(recv_buf, rc);
+				VERIFY(ibuf_alloc(recv_buf, rc) != NULL);
 			} else if (rc == IOSTREAM_ERROR) {
 				goto io_error;
 			} else {
diff --git a/src/lib/core/coio_buf.h b/src/lib/core/coio_buf.h
index 908a32c7e..110cc700c 100644
--- a/src/lib/core/coio_buf.h
+++ b/src/lib/core/coio_buf.h
@@ -50,7 +50,7 @@ coio_bread(struct iostream *io, struct ibuf *buf, size_t sz)
 	ssize_t n = coio_read_ahead(io, buf->wpos, sz, ibuf_unused(buf));
 	if (n < 0)
 		diag_raise();
-	ibuf_alloc(buf, n);
+	VERIFY(ibuf_alloc(buf, n) != NULL);
 	return n;
 }
 
@@ -68,7 +68,7 @@ coio_bread_timeout(struct iostream *io, struct ibuf *buf, size_t sz,
 			                    timeout);
 	if (n < 0)
 		diag_raise();
-	ibuf_alloc(buf, n);
+	VERIFY(ibuf_alloc(buf, n) != NULL);
 	return n;
 }
 
@@ -80,7 +80,7 @@ coio_breadn(struct iostream *io, struct ibuf *buf, size_t sz)
 	ssize_t n = coio_readn_ahead(io, buf->wpos, sz, ibuf_unused(buf));
 	if (n < 0)
 		diag_raise();
-	ibuf_alloc(buf, n);
+	VERIFY(ibuf_alloc(buf, n) != NULL);
 	return n;
 }
 
@@ -99,7 +99,7 @@ coio_breadn_timeout(struct iostream *io, struct ibuf *buf, size_t sz,
 			                     timeout);
 	if (n < 0)
 		diag_raise();
-	ibuf_alloc(buf, n);
+	VERIFY(ibuf_alloc(buf, n) != NULL);
 	return n;
 }
 
```

</details>

Comment (nshy) [reply src/box/lua/net_box.c]: Done. See new commits:
 - `"misc: use consume for read end of the ibuf"`
 - `"iproto: clean accounting of processed messages in input buffers"`

Changes depend of PR https://github.com/tarantool/small/pull/82

File: src/box/lua/merger.c [hunk 3]
```diff
@@ -115,8 +115,8 @@ decode_header(struct ibuf *buf, size_t *len_p)
 static void
 encode_header(struct ibuf *output_buffer, uint32_t result_len)
 {
-	ibuf_reserve(output_buffer, mp_sizeof_array(result_len));
-	output_buffer->wpos = mp_encode_array(output_buffer->wpos, result_len);
+	mp_encode_array(xibuf_alloc(output_buffer, mp_sizeof_array(result_len)),
```
Comment (locker) [inline src/box/lua/merger.c:118]: Nit: Let's use a variable:

```C
void *wpos = xibuf_alloc(...);
mp_encode_array(wpos, ...);
```

Comment (nshy) [reply src/box/lua/merger.c:118]: Done.

```diff
diff --git a/src/box/lua/merger.c b/src/box/lua/merger.c
index 20f6be1dc..c17671c2f 100644
--- a/src/box/lua/merger.c
+++ b/src/box/lua/merger.c
@@ -115,8 +115,8 @@ decode_header(struct ibuf *buf, size_t *len_p)
 static void
 encode_header(struct ibuf *output_buffer, uint32_t result_len)
 {
-       mp_encode_array(xibuf_alloc(output_buffer, mp_sizeof_array(result_len)),
-                       result_len);
+       void *wpos = xibuf_alloc(output_buffer, mp_sizeof_array(result_len));
+       mp_encode_array(wpos, result_len);
 }

 /**
```

File: src/box/lua/merger.c [hunk 4]
```diff
@@ -596,10 +596,10 @@ luaL_merge_source_buffer_next(struct merge_source *base,
 		return -1;
 	}
 	--source->remaining_tuple_count;
-	source->buf->rpos = (char *) tuple_end;
 	if (format == NULL)
 		format = tuple_format_runtime;
 	struct tuple *tuple = tuple_new(format, tuple_beg, tuple_end);
+	ibuf_consume_before(source->buf, (char *)tuple_end);
```
Comment (locker) [inline src/box/lua/merger.c:602]: The cast isn't needed (I changed the argument type to `const void *`). Please update the submodule and remove the cast.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/lua/merger.c (hunk 4 above)
Comment (nshy) [reply src/box/lua/merger.c:602]: Done.

```diff
diff --git a/src/box/lua/merger.c b/src/box/lua/merger.c
index c17671c2f..99c5b31b1 100644
--- a/src/box/lua/merger.c
+++ b/src/box/lua/merger.c
@@ -599,7 +599,7 @@ luaL_merge_source_buffer_next(struct merge_source *base,
        if (format == NULL)
                format = tuple_format_runtime;
        struct tuple *tuple = tuple_new(format, tuple_beg, tuple_end);
-       ibuf_consume_before(source->buf, (char *)tuple_end);
+       ibuf_consume_before(source->buf, tuple_end);
        if (tuple == NULL)
                return -1;
```

**Review by locker — APPROVED** [review]

**Pull request merged**